### PR TITLE
expression: rewrite builtin function: DIV

### PR DESCRIPTION
**Pull request by hanfei1991**

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -40,6 +41,8 @@ var (
 	_ builtinFunc = &builtinArithmeticMinusDecimalSig{}
 	_ builtinFunc = &builtinArithmeticMinusIntUnsignedSig{}
 	_ builtinFunc = &builtinArithmeticMinusIntSig{}
+	_ builtinFunc = &builtinArithmeticDivideRealSig{}
+	_ builtinFunc = &builtinArithmeticDivideDecimalSig{}
```
Comment (jackysp) [inline expression/builtin_arithmetic.go]: Division only supports real and decimal?

File: expression/builtin_arithmetic.go [hunk 2]
```diff
@@ -453,6 +456,77 @@ func (s *builtinArithmeticMultiplyIntSig) evalInt(row []types.Datum) (val int64,
...
+		return nil, errors.Trace(err)
+	}
+	tpA, tpB := args[0].GetType(), args[1].GetType()
+	tcA, tcB := numericContextResultType(tpA), numericContextResultType(tpB)
+	if tcA == types.ClassReal || tcB == types.ClassReal {
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpReal, tpReal, tpReal)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		setFlenDecimal4RealOrDecimal(bf.tp, tpA, tpB, true)
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: in TiDB:
```sql
TiDB > select 1.99 * 2.991;
Field   1:  `1.99 * 2.991`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     8
Max_length: 7
Decimals:   3
Flags:      BINARY NUM


+--------------+
| 1.99 * 2.991 |
+--------------+
|      5.95209 |
+--------------+
1 row in set (0.00 sec)
```

in MySQL:
```sql
MySQL > select 1.99 * 2.991;
Field   1:  `1.99 * 2.991`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     9
Max_length: 7
Decimals:   5
Flags:      NOT_NULL BINARY NUM


+--------------+
| 1.99 * 2.991 |
+--------------+
|      5.95209 |
+--------------+
1 row in set (0.00 sec)
```

`Flen` and `Decimal` can not be set using `setFlenDecimal4RealOrDecimal` for builtin function div and mul

**Comment by zz-jason** [comment]
@hanfei1991 ci failed.

**Comment by hanfei1991** [comment]
@zz-jason @XuHuaiyu PTAL

File: expression/builtin_arithmetic.go [hunk 3]
```diff
@@ -453,6 +485,77 @@ func (s *builtinArithmeticMultiplyIntSig) evalInt(row []types.Datum) (val int64,
...
+		}
+		setType4DivReal(bf.tp)
+		sig := &builtinArithmeticDivideRealSig{baseRealBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpDecimal, tpDecimal, tpDecimal)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	setType4DivDecimal(bf.tp, tpA, tpB)
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: `Flen` and `Decimal` is not correct in `arithmeticMultiplyFunctionClass`, do `MUL` and `DIV` share the same rule to inference `Flen` and `Decimal` ?

Comment (hanfei1991) [inline expression/builtin_arithmetic.go]: The rules are different. I will fix it in the future.

**Comment by XuHuaiyu** [comment]
LGTM

File: expression/builtin_arithmetic.go [hunk 4]
```diff
@@ -40,13 +41,19 @@ var (
...
 	_ builtinFunc = &builtinArithmeticMultiplyRealSig{}
 	_ builtinFunc = &builtinArithmeticMultiplyDecimalSig{}
 	_ builtinFunc = &builtinArithmeticMultiplyIntUnsignedSig{}
 	_ builtinFunc = &builtinArithmeticMultiplyIntSig{}
 	_ builtinFunc = &builtinArithmeticSig{}
 )
 
+// precIncrement indicates the number of digits by which to increase the scale of the result of division operations
+// performed with the / operator.
+const precIncrement = 4
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: if it's only used in type inference of `DIV`, it's better to be a member of `arithmeticDivideFunctionClass`

File: expression/builtin_arithmetic.go [hunk 5]
```diff
@@ -91,6 +98,31 @@ func setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
 	retTp.Flen = types.UnspecifiedLength
 }
 
+func setType4DivDecimal(retTp, a, b *types.FieldType) {
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: ditto

File: expression/builtin_arithmetic.go [hunk 6]
```diff
@@ -91,6 +98,31 @@ func setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
...
+	}
+	if a.Flen == types.UnspecifiedLength {
+		retTp.Flen = types.UnspecifiedLength
+		return
+	}
+	retTp.Flen = a.Flen + decb + precIncrement
+	retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxDecimalWidth)))
+}
+
+func setType4DivReal(retTp *types.FieldType) {
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: ditto

File: expression/builtin_arithmetic.go (hunk 4 above)
Comment (hanfei1991) [inline expression/builtin_arithmetic.go]: It is a session variable in mysql. I will change it in the future.

**Pull request merged**